File: egraph-js/layout/fm3.js
import {getModule} from '..'

export class FM3 {
  constructor () {
    this.functions = getModule().functions
    this.pointer = this.functions.layoutFM3New()
  }

  call (graph) {
    const start = Date.now()
    this.functions.layoutFM3Call(this.pointer, graph.pointer)
    const stop = Date.now()
    console.log(stop - start)
  }

  get minSize () {
    return this.functions.layoutFM3GetMinSize(this.pointer)
  }

  set minSize (value) {
    this.functions.layoutFM3SetMinSize(this.pointer, value)
  }

  get stepIteration () {
    return this.functions.layoutFM3GetStepIteration(this.pointer)
  }

  set stepIteration (value) {
    this.functions.layoutFM3SetStepIteration(this.pointer, value)
  }

  get unitEdgeLength () {
    return this.functions.layoutFM3GetUnitEdgeLength(this.pointer)
  }

  set unitEdgeLength (value) {
    this.functions.layoutFM3SetUnitEdgeLength(this.pointer, value)
  }

  get manyBodyForceStrength () {
    return this.functions.layoutFM3GetManyBodyForceStrength(this.pointer)
  }

  set manyBodyForceStrength (value) {
    this.functions.layoutFM3SetManyBodyForceStrength(this.pointer, value)
  }

  get linkForceStrength () {
    return this.functions.layoutFM3GetLinkForceStrength(this.pointer)
  }

  set linkForceStrength (value) {
    this.functions.layoutFM3SetLinkForceStrength(this.pointer, value)
  }

  get positionForceStrength () {
    return this.functions.layoutFM3GetPositionForceStrength(this.pointer)
  }

  set positionForceStrength (value) {
    this.functions.layoutFM3SetPositionForceStrength(this.pointer, value)
  }
}

File: egraph-c-api/src/layout/fm3.rs
use std::os::raw::{c_double, c_uint};
use egraph::layout::fm3::FM3;
use ::graph::Graph;

#[no_mangle]
pub unsafe fn layout_fm3_new() -> *mut FM3 {
    let fm3 = Box::new(FM3::new());
    Box::into_raw(fm3)
}

#[no_mangle]
pub unsafe fn layout_fm3_get_min_size(p_fm3: *mut FM3) -> c_uint {
    (*p_fm3).min_size as c_uint
}

#[no_mangle]
pub unsafe fn layout_fm3_set_min_size(p_fm3: *mut FM3, value: c_uint) {
    (*p_fm3).min_size = value as usize;
}

#[no_mangle]
pub unsafe fn layout_fm3_get_step_iteration(p_fm3: *mut FM3) -> c_uint {
    (*p_fm3).step_iteration as c_uint
}

#[no_mangle]
pub unsafe fn layout_fm3_set_step_iteration(p_fm3: *mut FM3, value: c_uint) {
    (*p_fm3).step_iteration = value as usize;
}

#[no_mangle]
pub unsafe fn layout_fm3_get_unit_edge_length(p_fm3: *mut FM3) -> c_double {
    (*p_fm3).unit_edge_length as c_double
}

#[no_mangle]
pub unsafe fn layout_fm3_set_unit_edge_length(p_fm3: *mut FM3, value: c_double) {
    (*p_fm3).unit_edge_length = value as f32;
}

#[no_mangle]
pub unsafe fn layout_fm3_call(p_fm3: *mut FM3, p_graph: *mut Graph) {
    let points = (*p_fm3).call(&*p_graph);
    for (node, point) in (*p_graph).node_indices().zip(points) {
        let mut node = (*p_graph).node_weight_mut(node).unwrap();
        node.x = point.x as f64;
        node.y = point.y as f64;
    }
}
